includes/class-wp-document-revisions-validate-structure.php

```diff
@@ -0,0 +1,628 @@
+			 AND post_status not in ( 'auto-draft', 'trash' )
+			 ORDER BY ID DESC
+			",
+			ARRAY_A
+		);
+		// phpcs:enable WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.PreparedSQL.NotPrepared,WordPress.DB.DirectDatabaseQuery
+		$num_doc  = $wpdb->num_rows;
+		$failures = array();
+		foreach ( $documents as $rec => $document ) {
+			// check that user can edit the document.
+			if ( current_user_can( 'edit_document', $document['ID'] ) ) {
+				$test = self::validate_document( $document['ID'], $document['post_content'], $document['post_modified_gmt'] );
+				if ( is_array( $test ) ) {
+					// failure.
+					$failures[] = array_merge( $document, $test );
+				}
+			}
+		}
+		// No errors found.
+		if ( empty( $failures ) ) {
+			echo '<h2 class="title">' . esc_html__( 'Invalid Document Internal Structures', 'wp-document-revisions' ) . '</h2>';
+			echo '<p>' . esc_html__( 'No invalid documents found.', 'wp-document-revisions' ) . '</p>';
+			return;
+		}
+		?>
+		<div class="wrap">
+		<h2 class="title"><?php esc_html_e( 'Invalid Document Internal Structures', 'wp-document-revisions' ); ?></h2>
+
+			<div id="col-container">
+				<table class="widefat" cellspacing="0">
+					<thead>
+						<tr>
+							<th scope="col" id="label" class="manage-column column-name"><?php esc_html_e( 'ID', 'wp-document-revisions' ); ?></th>
+							<th scope="col" id="label" class="manage-column column-name"><?php esc_html_e( 'Type', 'wp-document-revisions' ); ?></th>
+							<th scope="col" id="label" class="manage-column column-name"><?php esc_html_e( 'Title', 'wp-document-revisions' ); ?></th>
+							<th scope="col" id="name"  class="manage-column column-name"><?php esc_html_e( 'Message', 'wp-document-revisions' ); ?></th>
+							<th scope="col" id="slug"  class="manage-column column-name"><?php esc_html_e( 'Fix', 'wp-document-revisions' ); ?></th>
+						</tr>
+					</thead>
+
+					<tbody id="the-list" class="list:failures">
+					<?php
+					foreach ( $failures as $rec => $failure ) {
+						$line = esc_attr( $failure['ID'] );
+						$beg  = esc_attr( $failure['post_title'] ) . '<br/>';
+						$ref  = esc_url( get_the_permalink( $failure['ID'] ) );
+						if ( (bool) $failure['fix'] ) {
+							if ( (bool) $failure['error'] ) {
+								$ref = $beg . '<div id="on_' . $line . '" style="display: block;">' . $ref . '</div><div id="off' . $line .
+									'" style="display: none;"><a href="' . $ref . '" target="_blank">' . $ref . '</a></div>';
+							} else {
+								$ref = $beg . '<a href="' . $ref . '" target="_blank">' . $ref . '</a>';
+							}
+						} else {
+							$ref = $beg . $ref;
+						}
+						if ( (bool) $failure['fix'] ) {
+							$fix = '<button onclick="wpdr_valid_fix(' . $failure['ID'] . ',' . $failure['code'] . ',' . $failure['parm'] . ')">' . esc_html__( 'Fix', 'wp-document-revisions' ) . '</button>';
+						} else {
+							$fix = '';
+						}
+						// phpcs:disable  WordPress.Security.EscapeOutput
```

Comment:
To avoid the security warning, could this be rewritten with inline HTML between the conditionals + string interpolation, instead of building the HTML string in PHP and outputting it all at once?

Reply:
I think that this is always a trade-off. I don't intrinsically think that there is a problem to put a warning exclusion (even though it allowed me to let a potential error through). 

For me the problem here was that with so many bits and pieces going into making up the URL if it went into the row definition, it woulf be difficult to see the pattern. Also looking at the variable $line (which is the document ID, ie. an internal number). It is used 5 times. So wrapping it 5 times in esc_attr(  ) seems excessive.

I have added some comments to what the cell contents are.